Support ordered case-insensitive duration strings for timedelta

ddelange opened the pull request:
Duration strings like `7m7s` are widely used for configs, especially anything kubernetes related (anything go-related really, because of the `time.ParseDuration` builtin).

This PR adds a simple yet robust regex parser for such strings.

For timedelta specifically, this also incorporates #297

sloria commented:
Thanks for the PR! This is interesting. Do you have a source from which you derived the regex?

I'm open to this feature, but it'd be good to have a spec or a source that we can refer to if we need to update it.

ddelange commented:
Hi @sloria :wave:

Thanks for the quick reply!
- The go ParseDuration implementation has “ns”, “us” (or “µs”), “ms”, “s”, “m”, “h”.
- ISO 8601 time interval looks like P3Y6M4DT12H30M5S (P for period, and then partial overlap with go).
- Python timedelta has a subset of both, plus `weeks`.

The PR now supports all units supported by Python timedelta. I wrote the pattern myself and added the comments for maintainability.

sloria commented:
Thanks, that's very helpful info 👍 . Above the regex, can you add a link to the spec and add a comment block with the discrepancies you listed?

sloria commented on CHANGELOG.md:
```diff
@@ -1,6 +1,6 @@
 # Changelog
 
-## 11.1.0 (2025-11-11)
```
good catch

ddelange commented:
@sloria 
- docs added
- had another look at the spec and added support for `0s` `0h` etc
- had another look at the marshmallow implementation and added support for `-42s` etc
- expanded tests for these cases, added some expected failures

sloria commented:
Thank you! Looks great!

Is there a good reason to make the regex case-insensitive? It could be more future-proof to keep it case-sensitive. Uppercase characters are probably a typo by the user anyway, right?

ddelange commented:
Sure thing, done 💥

sloria commented:
Thank you! This looks good to go

sloria commented:
Released in 11.2.0

ddelange commented:
awesome, thanks for the collab and for quickly landing this!
![landed](https://feross.net/x/cats/29.gif)

ddelange commented:
interesting, marshmallow expects a int for timedelta, and this PR switched to subclassing TimeDelta. maybe there was some lenience that was silently swallowing the float
